src/compile/scale.ts

```diff
@@ -191,79 +292,114 @@ export function type(specifiedType: ScaleType, fieldDef: FieldDef, channel: Chan
-  if (contains([ROW, COLUMN], channel)) {
-    return ScaleType.BAND;
-  }
-  if (channel === SHAPE) {
-    return ScaleType.ORDINAL_LOOKUP;
+namespace defaultProperty {
+  /**
+   * Determine appropriate default scale type.
+   */
+  export function type(fieldDef: FieldDef, channel: Channel, mark: Mark, canHaveBandSize: boolean): ScaleType {
+    if (contains([ROW, COLUMN], channel)) {
+      return ScaleType.BAND;
+    }
+    if (channel === SHAPE) {
+      return ScaleType.ORDINAL_LOOKUP;
+    }
+
+    switch (fieldDef.type) {
+      case NOMINAL:
+        if (channel === COLOR) {
+          return ScaleType.ORDINAL_LOOKUP;
+        }
+        return ordinalToContinuousType(channel, mark, canHaveBandSize);
+      case ORDINAL:
+        if (channel === COLOR) {
+          // TODO: check if this is still true
+          return ScaleType.LINEAR; // ordinal has order, so use interpolated ordinal color scale.
+        }
+        return ordinalToContinuousType(channel, mark, canHaveBandSize);
+      case TEMPORAL:
+        if (channel === COLOR) {
+          // FIXME: or sequential?
+          return ScaleType.TIME; // time has order, so use interpolated color scale.
+        }
+
+        switch (fieldDef.timeUnit) {
+          // These time unit use discrete scale by default
+          case TimeUnit.HOURS:
+          case TimeUnit.DAY:
+          case TimeUnit.MONTH:
+          case TimeUnit.QUARTER:
+            return ordinalToContinuousType(channel, mark, canHaveBandSize);
+        }
+        return ScaleType.TIME;
+
+      case QUANTITATIVE:
+        return ScaleType.LINEAR;
+    }
+
+    // should never reach this
+    return null;
   }
 
-  switch (fieldDef.type) {
-    case NOMINAL:
-      if (channel === COLOR) {
-        return ScaleType.ORDINAL_LOOKUP;
-      }
-      return ordinalToContinuousType(channel, mark, canHaveBandSize);
-    case ORDINAL:
-      if (channel === COLOR) {
-        // TODO: check if this is still true
-        return ScaleType.LINEAR; // ordinal has order, so use interpolated ordinal color scale.
-      }
-      return ordinalToContinuousType(channel, mark, canHaveBandSize);
-    case TEMPORAL:
-      if (channel === COLOR) {
-        // FIXME: or sequential?
-        return ScaleType.TIME; // time has order, so use interpolated color scale.
+  /**
+   * @returns BAND or POINT scale based on channel, mark, and bandSize
+   */
+  function ordinalToContinuousType(channel: Channel, mark: Mark, canHaveBandSize: boolean): ScaleType {
+    if (contains([X, Y], channel)) {
+      // Use band ordinal scale for x/y scale in one of the following cases:
+      if (
+        // 1) the mark is bar and the scale's bandWidth is 'fit',
+        (mark === BAR && !canHaveBandSize) ||
+        // 2) the mark is rect
+        mark === RECT
+      ) {
+        return ScaleType.BAND;
       }
+    }
+    // Otherwise use ordinal point scale
```

Comment:
I don't like these comments that just explain code. Better say why something is done the way it is done.